EC-4013 - Add "cacheMode" parameter properly

radu-todirica opened the pull request:
EC-4013 - Add "cacheMode" parameter properly
EC-3865 - Update disk and service offering endpoints

## Description
<!--- Describe your changes in detail -->

Currently the createDiskOffering endpoint in our version of cloudstack has a "cacheMode" parameter that is not documented properly and therefore not returned as part of the listApis response.

It also isn't present on the Service offering creation endpoint.

So, this is what needs to happen:

1. "cacheMode" field need to be added to the createServiceOffering endpoint
2. the "cacheMode" field needs to be returned in the listApis response for both createDiskOffering and createServiceOffering endpoints.
3. the "cacheMode" field needs to be returned in the listServiceOfferings and listDiskOfferings endpoints.

Also this PR contains the changes for the story EC-3865 - Update disk and service offering endpoints in order to be able to update any field without passing in un-changed fields.

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document -->

radu-todirica committed:
EC-4013 - Add "cacheMode" parameter properly
EC-3865 - Update disk and service offering endpoints

nathanejohnson commented on api/src/org/apache/cloudstack/api/command/admin/offering/CreateDiskOfferingCmd.java:
```diff
@@ -130,6 +130,9 @@
+    @Parameter(name = ApiConstants.CACHE_MODE, type = CommandType.STRING, required = true ,description = "the cache mode to use for this disk offering. none, writeback or writethrough")
+    private String cacheMode;
+
 /////////////////////////////////////////////////////
```
could this be an enum type?

radu-todirica replied:
Done. I changed the cache mode to an enum.

nathanejohnson reviewed (changes requested):
Overall this looks good but can we make cache mode an enum?

radu-todirica committed:
EC-4013 - Add "cacheMode" parameter properly
EC-3865 - Update disk and service offering endpoints
EC-4014 - Disk offering cache mode not used when attaching volume to VM

radu-todirica committed:
Revert "EC-4013 - Add "cacheMode" parameter properly"

This reverts commit 199156d8bcac53a19d302155b3182d1b4dccfc88.

radu-todirica committed:
fixup! Revert "EC-4013 - Add "cacheMode" parameter properly"

radu-todirica committed:
Add a validation on cache mode parameter

radu-todirica committed:
Add a validation fix on cache mode parameter

nathanejohnson commented on plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:
```diff
@@ -1092,7 +1095,7 @@ public Answer attachVolume(final AttachCommand cmd) {
                     vol.getBytesWriteRate(), vol.getBytesWriteRateMax(), vol.getBytesWriteRateMaxLength(),
                     vol.getIopsReadRate(), vol.getIopsReadRateMax(), vol.getIopsReadRateMaxLength(),
-                    vol.getIopsWriteRate(), vol.getIopsWriteRateMax(), vol.getIopsWriteRateMaxLength());
+                    vol.getIopsWriteRate(), vol.getIopsWriteRateMax(), vol.getIopsWriteRateMaxLength(), vol.getCacheMode().toString());
```
if cache mode is null, this will throw an exception

nathanejohnson commented on plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:
```diff
@@ -1121,7 +1124,7 @@ public Answer dettachVolume(final DettachCommand cmd) {
                     vol.getBytesWriteRate(), vol.getBytesWriteRateMax(), vol.getBytesWriteRateMaxLength(),
                     vol.getIopsReadRate(), vol.getIopsReadRateMax(), vol.getIopsReadRateMaxLength(),
-                    vol.getIopsWriteRate(), vol.getIopsWriteRateMax(), vol.getIopsWriteRateMaxLength());
+                    vol.getIopsWriteRate(), vol.getIopsWriteRateMax(), vol.getIopsWriteRateMaxLength(), vol.getCacheMode().toString());
```
if cache mode is null, this will throw an exception

radu-todirica committed:
Bug EC-4123 Attach/detach disk is failing.

radu-todirica committed:
Changed the cache_mode parameter required to false.

radu-todirica committed:
Add a validation against cache_mode on create service method.
